src/notation/internal/masternotationmididata.cpp

```diff
@@ -0,0 +1,576 @@
+
+        std::list<InstrumentChannel*> channelList(instrument->channel().begin(), instrument->channel().end());
+
+        std::vector<Event> setupEvents = retrieveSetupEvents(channelList);
+        stream.controlEventsStream.set(std::move(setupEvents));
+
+        stream.eventsRequest.onReceive(this, [this, stream, midiChannels](const tick_t fromTick, const tick_t toTick) mutable {
+            if (fromTick >= stream.lastTick || toTick > stream.lastTick) {
+                stream.mainStream.send({}, 0);
+                return;
+            }
+
+            for (const channel_t& midiChannel : midiChannels) {
+                Events events = retrieveEvents(static_cast<channel_t>(midiChannel), fromTick, toTick);
+
+                stream.mainStream.send(std::move(events), toTick);
+            }
+        });
+    }
+
+    return stream;
+}
+
+TempoMap MasterNotationMidiData::makeTempoMap() const
+{
+    midi::TempoMap tempos;
+
+    Ms::TempoMap* tempomap = score()->tempomap();
+    qreal relTempo = tempomap->relTempo();
+    for (const Ms::RepeatSegment* rs : score()->repeatList()) {
+        int startTick = rs->tick, endTick = startTick + rs->len();
+        int tickOffset = rs->utick - rs->tick;
+
+        auto se = tempomap->lower_bound(startTick);
+        auto ee = tempomap->lower_bound(endTick);
+        for (auto it = se; it != ee; ++it) {
+            //
+            // compute midi tempo: microseconds / quarter note
+            //
+            tempo_t tempo = static_cast<tempo_t>(lrint((1.0 / (it->second.tempo * relTempo)) * 1000000.0));
+
+            tempos.insert({ it->first + tickOffset, tempo });
+        }
+    }
+
+    return tempos;
+}
+
+Ret MasterNotationMidiData::playNoteMidiData(const Ms::Note* note) const
+{
+    const Ms::Note* masterNote = note;
+    if (note->linkList().size() > 1) {
+        for (Ms::ScoreElement* se : note->linkList()) {
+            if (se->score() == note->masterScore() && se->isNote()) {
+                masterNote = Ms::toNote(se);
+                break;
+            }
+        }
+    }
+
+    Ms::Fraction tick = masterNote->chord()->tick();
+    if (tick < Ms::Fraction(0, 1)) {
+        tick = Ms::Fraction(0, 1);
+    }
+    const Ms::Instrument* instr = masterNote->part()->instrument(tick);
+    channel_t midiChannel = instr->channel(masterNote->subchannel())->channel();
+
+    MidiData midiData = trackMidiData(masterNote->part()->id());
+
+    Events events = retrieveEventsForElement(masterNote, midiChannel);
+    midiData.stream.backgroundStream.send(std::move(events), Ms::MScore::defaultPlayDuration* 2);
+
+    return make_ret(Ret::Code::Ok);
+}
+
+Ret MasterNotationMidiData::playChordMidiData(const Ms::Chord* chord) const
+{
+    Ms::Part* part = chord->part();
+    Ms::Fraction tick = chord->segment() ? chord->segment()->tick() : Ms::Fraction(0, 1);
+    Ms::Instrument* instr = part->instrument(tick);
+    channel_t midiChannel = instr->channel(chord->notes()[0]->subchannel())->channel();
```

Comment:
Please, add a check on emptiness here

Reply:
well, if there is a chord without any note, than it should be an exceptional case
added IF_ASSERT statement